### PR TITLE
Generate new deployment for each app version

### PR DESCRIPTION
**Pull request by rgrandl**
In the current implementation, we generate new deployments for each replica set other than the one that has the public listeners (e.g., main). When we rollout a new version of the app, we do rolling upgrade on the main group, hence we can do automated rollouts.

However, this approach is not great and it doesn't make sense for someone to do this in production. For development/testing, someone can simply start a new app deployment and kill the old one.

This PR changes the deployment such that:
* we now generate unique deployments for all groups (including main)
* we generate unique name for the listener service iff the user wants that; this allows the user to reuse the same listener across multiple app versions - makes it easy to do port-forwarding, curl the same IP, etc.
* we added a "version" label to the deployments/HPAs/listeners, s.t., we can easily delete old deployments; e.g., `kubectl delete all --selector=version=foo` deletes all the k8s resources for the app version foo.

### CODE REVIEW

File: internal/impl/kube.go [hunk 1]
```diff
@@ -268,16 +262,17 @@ func (r *replicaSet) buildListenerService(lis *ReplicaSetConfig_Listener) (*core
 			Kind:       "Service",
 		},
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      globalLisName,
+			Name:      lisServiceName,
 			Namespace: r.namespace,
 			Labels: map[string]string{
 				"lisName": lis.Name,
+				"version": version,
```
Comment (mwhittaker) [inline internal/impl/kube.go:269]: I think the version should always be `r.dep.Id[:8]`, even if the user specifies a service name?

File: internal/impl/kube.go [hunk 2]
```diff
@@ -443,6 +423,8 @@ func GenerateYAMLs(image string, dep *protos.Deployment, cfg *KubeConfig) error
 
 	// Generate roles and role bindings.
 	var generated []byte
+	header := fmt.Sprintf("# To delete this deployment run:\n#  `kubectl delete all --selector=version=%s`\n\n", dep.Id[:8])
```
Comment (mwhittaker) [inline internal/impl/kube.go:426]: Love it!

File: internal/impl/kube.go [hunk 3]
```diff
@@ -234,10 +224,6 @@ func (r *replicaSet) buildDeployment(cfg *KubeConfig) (*appsv1.Deployment, error
 					HostNetwork: cfg.UseHostNetwork,
 				},
 			},
-			Strategy: appsv1.DeploymentStrategy{
-				Type:          "RollingUpdate",
-				RollingUpdate: &appsv1.RollingUpdateDeployment{},
-			},
 			// Number of old ReplicaSets to retain to allow rollback.
 			RevisionHistoryLimit: ptrOf(int32(1)),
```
Comment (mwhittaker) [inline internal/impl/kube.go:228]: Can we delete `RevisionHistoryLimit` and `MinReadySeconds` now too?

**Review by mwhittaker — APPROVED** [review]
Nice!!!

File: internal/impl/kube.go (hunk 3 above)
Comment (rgrandl) [reply internal/impl/kube.go]: Good point about 'RevisionHistoryLimit`. As for `MinReadySeconds` I think it still makes sense, but I guess we can let someone configure it if they really want so we set less knobs in the config.

**Review by rgrandl — COMMENTED** [review]
Thanks Michael

**Pull request merged**